Userland/Applications/Screenshot/ScreenshotWidget.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2021, the SerenityOS Developers
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "ScreenshotWidget.h"
+#include "SelectableOverlay.h"
+#include <Applications/Screenshot/ScreenshotWindowGML.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/DateTime.h>
+#include <LibCore/File.h>
+#include <LibCore/StandardPaths.h>
+#include <LibCore/Timer.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/Button.h>
+#include <LibGUI/CheckBox.h>
+#include <LibGUI/Clipboard.h>
+#include <LibGUI/Label.h>
+#include <LibGUI/Notification.h>
+#include <LibGUI/RadioButton.h>
+#include <LibGUI/SpinBox.h>
+#include <LibGUI/WindowServerConnection.h>
+#include <LibGfx/PNGWriter.h>
+#include <LibConfig/Client.h>
+#include <serenity.h>
+
+ScreenshotWidget::ScreenshotWidget()
+{
+    load_from_gml(screenshot_window_gml);
+
+    m_whole_button = *find_descendant_of_type_named<GUI::RadioButton>("wholescreen");
+    m_custom_button = *find_descendant_of_type_named<GUI::RadioButton>("customregion");
+    m_delay_spinbox = *find_descendant_of_type_named<GUI::SpinBox>("delay");
+    m_copy_checkbox = *find_descendant_of_type_named<GUI::CheckBox>("copy_to_clipboard");
+    m_cancel_button = *find_descendant_of_type_named<GUI::Button>("cancel_button");
+    m_ok_button = *find_descendant_of_type_named<GUI::Button>("ok_button");
+
+    m_window_sel = GUI::Window::construct();
+    m_overlay = m_window_sel->set_main_widget<SelectableOverlay>(m_window_sel);
+    m_overlay->callback = [&] {
+        save_screenshot(true);
+    };
+
+    m_window_sel->set_title("Screenshot");
+    m_window_sel->set_has_alpha_channel(true);
+    m_window_sel->set_fullscreen(true);
+
+    String type = Config::read_string("Screenshot", "General", "ScreenshotType", "Whole");
+    if (type == "Custom") {
+        m_whole_button->set_checked(false);
+        m_custom_button->set_checked(true);
+    } else {
+        m_whole_button->set_checked(true);
+        m_custom_button->set_checked(false);
+    }
+
+    m_delay_spinbox->set_value(Config::read_i32("Screenshot", "General", "Delay", 0));
+    m_copy_checkbox->set_checked(Config::read_bool("Screenshot", "General", "CopyToClipboard", true));
+
+    m_cancel_button->on_click = [&](auto) {
+        GUI::Application::the()->quit();
+    };
+
+    m_ok_button->on_click = [&](auto) {
+        String new_type = "Whole";
+        if (m_custom_button->is_checked()) {
+            new_type = "Custom";
+        }
+
+        Config::write_string("Screenshot", "General", "ScreenshotType", new_type);
+        Config::write_i32("Screenshot", "General", "Delay", m_delay_spinbox->value());
+        Config::write_bool("Screenshot", "General", "CopyToClipboard", m_copy_checkbox->is_checked());
+
+        window()->close();
+
+        m_output_path = Core::DateTime::now().to_string("screenshot-%Y-%m-%d-%H-%M-%S.png");
+
+        usleep(m_delay_spinbox->value() * 1000);
```

Comment:
As discussed on Discord, we should not be blocking the core event loop. Aside from just not being a nice thing to do, it will make the kernel / WindowServer think the application is unresponsive.

Reply:
This has already been discussed on Discord. It should be perfectly fine to block it here since the main window gets closed beforehand. Also, using a Timer doesn't work. I might give it a try again tomorrow but for now it works perfectly and have never had any issues.

Reply:
I also tries pledging only the Screenshot domain but doesn't work as well since it still needs access to /tmp/portal/config. Don't ask me why though.

Reply:
You have to call pledge_domains before the very first call to unveil.

Reply:
> Also, using a Timer doesn't work
It does - I posted a gist on Discord showing how to use it with this app

Reply:
Aha! Didn't see it. Imma try it out.